src/irmin-pack/IO_intf.ml

```diff
@@ -16,27 +16,85 @@
 
 open! Import
 
+(** This interface defines a type [t] which wraps a file descriptor and provides some
+    additional support for metadata fields (version, offset), and support for periodic
+    flushing of data (writes to file go via an internal buffer which is flushed when the
+    amount of buffered data is beyond a certain limit). *)
 module type S = sig
   type t
+
   type path := string
 
   val v : version:Version.t option -> fresh:bool -> readonly:bool -> path -> t
+  (** [v ~version ~fresh ~readonly pth] creates a new IO instance. [version] must be
+      [Some] if the [pth] does not point to a file. If the file exists, then [fresh] will
+      cause all data to be discarded (except for metadata); if [fresh] is false, and the
+      version of the file is greater than [version], an exception is thrown. [readonly]
+      indicates that operations like {!append} which change the file data or metadata
+      should be disallowed. 
+
+      In order to store the metadata, a [header_size] of 16 bytes is reserved at the
+      beginning of the file; reads and writes to the file at offset [off] will be
+      translated to offset [off+header_size].  
+
+      For a given [pth], there should be at most one read/write instance. There can be
+      multiple readonly instances, but only one read/write instance is allowed. This is
+      not currently enforced (but could be, using lock files for example).  *)
+
   val name : t -> string
+  (** [name t] is the [path] argument provided to {!v} when [t] was created *)  
+
   val append : t -> string -> unit
+  (** [append t s] appends [s] to the end of the file pointed to by [t]. The data is first
+      placed in a buffer, and a flush to file only happens when the amount of data in the
+      buffer exceeds the [auto_flush_limit], which is currently 1M.
+
+      [append] and [set] are the only operations that write to the file, apart from
+      metadata operations. *)
+
   val set : t -> off:int63 -> string -> unit
+  (** [set t ~off s] writes the string [s] at offset [off+header_size]. *)
+
   val read : t -> off:int63 -> bytes -> int
+  (** [read t ~off buf] tries to read [Bytes.length buf] bytes from [t] at offset [off +
+      header_size]; it returns the number of bytes actually read, which may be less than
+      requested if we are near the end of the file *)
+
   val read_buffer : t -> off:int63 -> buf:bytes -> len:int -> int
+  (** [read_buffer t ~off ~buf ~len] tries to read [len] bytes from [t] at offset
+      [off+header_size] *)
+
   val offset : t -> int63
+  (** [offset t] returns the offset field of [t], which for the read/write instance is the
+      position at which data is written using {!append}. *)
+
   val force_offset : t -> int63
+  (** [force_offset t] returns the last offset for data that was reliably synced to disk
+      from the internal buffer, and as a side effect it updates the [t.offset] field to
```

Comment:
It's a little unclear to me what "offset" means here.

I the docs for `set` and `read` above I read "offset" to mean "real offset in the underlying file" (after adjusting for the header sizes), but `offset` and `force_offset` both return offsets in the contiguous data suffix built by `append`. I somewhat prefer the latter approach, as the `IO` abstraction exists partly to hide the details of metadata management & expose functions for interacting with the data itself, but either is fine so long as we are consistent.

Reply:
Are you saying we should change the code? I'm just aiming to document the code that is already there.

Reply:
I'm saying that the API needs to make it clear whether an "offset" is an offset into the underlying file or an offset into the data segment. The functions `offset` and `force_offset` return the latter, but the new doc-comments for `read` and `set` describe reading and writing at offset `off + header_size`. Perhaps those comments just need to say that they read/write from `off`-many bytes into the data segment?

Reply:
I'm still mystified about `force_offset`. Let's assume it is for readonly instances. Then what is it supposed to do? As I understand it, unlike Index, there is no need to "keep up with" data being written to a file. @CraigFe do you have any idea what this function is supposed to do?

Reply:
When a read-only instance explicitly `sync`s with the writer, it first checks if any additional data has been written to the pack file since the offset was last read ([here](https://github.com/mirage/irmin/blob/1d83f084815d11140f9b77fcb4732851bb6dec18/src/irmin-pack/pack_store.ml#L556-L558)). If so, then the dictionary and index are sync-ed too. `force_offset` exists in order to support this explicit sync process in the three places it's called (`atomic_write.ml`, `dict.ml`, and `pack_store.ml`).

Reply:
Ah. So it is used to trigger syncs for the dictionary and index in the readonly instance. OK. I understood this at one point but then forgot when focused on pack store itself. So, readonly instances maintain "IO.offset" as a way to track when the underlying data changes, and "force_offset" is used to read when offset has been changed by the RW instance? So IO.offset doesn't just return the length of the underlying file for RO instances, it returns the value of the underlying offset field, which may lag the real offset.

Reply:
I guess a follow up question would be: what if the file data changes but the offset doesn't (eg via `set`)? If RO instances use increases in offset to trigger updates, then updating via set seems risky in that it wouldn't trigger an update in the RO instances (but I would expect that maybe it would, depending on what the desired semantics should be).